library/core/src/iter/adapters/dedup.rs

```diff
@@ -0,0 +1,169 @@
+use crate::mem::swap;
+
+/// An iterator that removes all but the first of consecutive elements in a
+/// given iterator according to the [`PartialEq`] trait implementation.
+///
+/// This `struct` is created by the [`dedup`] method on [`Iterator`]. See its
+/// documentation for more.
+///
+/// [`dedup`]: Iterator::dedup
+/// [`Iterator`]: trait.Iterator.html
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone, Copy)]
+pub struct Dedup<I, T> {
+    inner: I,
+    last: Option<T>,
+}
+
+impl<I, T> Dedup<I, T> {
+    pub(crate) const fn new(inner: I) -> Self {
+        Self { inner, last: None }
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<I, T> Iterator for Dedup<I, T>
+where
+    I: Iterator<Item = T>,
+    T: PartialEq,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.last.is_none() {
```

Comment:
Do we have to check this on every iteration?

Reply:
No, you're right, this is only strictly necessary for the first iteration. I can move this into the constructor.